traits/traits_listener.py

```diff
@@ -341,7 +341,7 @@ def register ( self, new ):
         """
         # Make sure we actually have an object to set listeners on and that it
         # has not already been registered (cycle breaking):
-        if (new is None) or (new is Undefined) or (new in self.active):
+        if (new is None) or (new is Undefined) or (new in self.active) or not isinstance(new, HasTraits):
```

Comment:
One minor comment: since `Undefined` and `None` are not instances of `HasTraits`, this whole condition could be simplified to `not isinstance(new, HasTraits) or new in self.active`.